### PR TITLE
Implementation of histogram with sycl kernel

### PR DESCRIPTION
**Pull request by AlexanderKalistratov**
Implemention of histogram with sycl_kernel.
This PR adds generic histogram kernel which can be used in the future to implement other versions of histogram such as `bincount`, `histogram2d` and `histogramdd` or specialize kernel for special cases like uniform bins.

sycl kernel covers only specific datatype and usm memory types. Unsupported cases are covered by additional copy.

### CODE REVIEW

File: dpnp/dpnp_iface_histograms.py [hunk 1]
```diff
@@ -393,44 +454,69 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+
+    if a_bin_dtype is None or hist_dtype is None:
+        raise ValueError(
+            f"function '{histogram}' does not support input types "
+            f"({a.dtype}, {bin_edges.dtype}, {ntype}), "
+            "and the inputs could not be coerced to any "
+            "supported types"
+        )
+
+    a_casted = a.astype(a_bin_dtype, order="C", copy=False)
```
Comment (oleksandr-pavlyk) [inline dpnp/dpnp_iface_histograms.py:480]: Because `a` may be `usm_ndarray` (according to docstring), it is not safe to use `a.astype`, since `usm_ndarray` does not implement such a method. Instead `dpctl.tensor.astype` should be used, or support for such input types should not be advertised.

**Comment by oleksandr-pavlyk** [comment]
Quick validation via independent implementation:

```python
def histogram1d_impl_tensor(data : dpt.usm_ndarray, bins : dpt.usm_ndarray) -> dpt.usm_ndarray:
    assert data.ndim == 1 
    assert bins.ndim == 1
    assert bins.shape[0] > 1
    bin_idx = dpt.searchsorted(bins, data)
    _, c = dpt.unique_counts(dpt.sort(bin_idx))
    return c
```

```

In [22]: x = dpnp.random.randn(10**7).get_array()

In [23]: bins = dpnp.asarray([-10, -4, -3, -2, -1, -0.5, -0.25, 0, 0.25, 0.5, 1, 2, 3, 4, 6], dtype=x.dtype).get_array()

In [24]: %time c, _ = dpnp.histogram(dpnp.asarray(x), bins=dpnp.asarray(bins)); print(c)
[    284   13222  214243 1359725 1497540  927356  987069  987352  927886
 1498601 1358597  214704   13114     307]
CPU times: user 10.6 ms, sys: 10.4 ms, total: 21 ms
Wall time: 15.2 ms

In [25]: %time c = histogram1d_impl_tensor(x, bins=bins); print(c)
[    284   13222  214243 1359725 1497540  927356  987069  987352  927886
 1498601 1358597  214704   13114     307]
CPU times: user 711 ms, sys: 581 ms, total: 1.29 s
Wall time: 396 ms
```

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+#include <algorithm>
+#include <complex>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+// dpctl tensor headers
+#include "utils/type_dispatch.hpp"
```
Comment (oleksandr-pavlyk) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:35]:
```suggestion
#include "dpctl4pybind11.hpp"
#include "utils/type_dispatch.hpp"
```

File: dpnp/dpnp_iface_histograms.py (hunk 1 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: Done, replaced with `dpnp.astype(a, ...)`

**Comment by AlexanderKalistratov** [comment]
@antonwolfy

File: dpnp/dpnp_iface_histograms.py [hunk 3]
```diff
@@ -393,48 +454,77 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+    bin_edges_casted = dpnp.astype(
+        bin_edges, a_bin_dtype, order="C", copy=False
+    )
+    weights_casted = (
+        dpnp.astype(weights, hist_dtype, order="C", copy=False)
+        if weights is not None
+        else None
+    )
+
+    n_usm_type = "device" if usm_type == "host" else usm_type
```
Comment (vtavana) [inline dpnp/dpnp_iface_histograms.py:490]: Why do we need change `usm_type` in this way?

File: dpnp/dpnp_iface_histograms.py [hunk 4]
```diff
@@ -306,6 +307,66 @@ def digitize(x, bins, right=False):
...
+    for t in supported:
+        if dpnp.can_cast(dt, t):
+            return t
+
+    return None
+
+
+def _result_type(dtype1, dtype2, has_fp64):
+    rt = dpnp.result_type(dtype1, dtype2)
+    if rt == dpnp.float64 and not has_fp64:
```
Comment (vtavana) [inline dpnp/dpnp_iface_histograms.py:323]: There is `map_dtype_to_device` function available that can be used for such cases. Using it has this benefit that the coverage report, will not show any missing line as it does currently.

`dpnp/dpnp_iface_histograms.py, coverage:   93%  missing lines: 213, 318, 324, 327, 336->339, 358, 359->362, 364-365, 473`

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+namespace dpctl_td_ns = dpctl::tensor::type_dispatch;
+using dpctl::tensor::usm_ndarray;
+
+using namespace histogram;
+
+namespace
+{
+
+template <typename T, typename BinsT, typename HistType = size_t>
+static sycl::event histogram_impl(sycl::queue exec_q,
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:53]:
```suggestion
static sycl::event histogram_impl(sycl::queue &exec_q,
```

File: dpnp/backend/extensions/sycl_ext/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+set(python_module_name _sycl_ext_impl)
+set(_module_src
+    ${CMAKE_CURRENT_SOURCE_DIR}/histogram.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/histogram_common.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/sycl_ext_py.cpp
+)
+
+pybind11_add_module(${python_module_name} MODULE ${_module_src})
+add_sycl_to_target(TARGET ${python_module_name} SOURCES ${_module_src})
+
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/CMakeLists.txt:36]: need to add for CUDA support:
```suggestion

if(_dpnp_sycl_targets)
    # make fat binary
    target_compile_options(
        ${python_module_name}
        PRIVATE
        -fsycl-targets=${_dpnp_sycl_targets}
    )
    target_link_options(
        ${python_module_name}
        PRIVATE
        -fsycl-targets=${_dpnp_sycl_targets}
    )
endif()

```

File: dpnp/backend/extensions/sycl_ext/histogram_common.hpp [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+{
+
+template <typename N, typename D>
+N CeilDiv(N n, D d)
+{
+    return (n + d - 1) / d;
+}
+
+template <typename N, typename D>
+N Align(N n, D d)
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram_common.hpp:52]: I believe it would be helpful to use `constexpr` in declaration:
```suggestion
constexpr N CeilDiv(N n, D d)
{
    return (n + d - 1) / d;
}

template <typename N, typename D>
constexpr N Align(N n, D d)
```

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    if (weights.has_value()) {
+        args_ev = dpctl::utils::keep_args_alive(
+            exec_q, {sample, bins, weights.value(), histogram}, {ev});
+    }
+    else {
+        args_ev = dpctl::utils::keep_args_alive(
+            exec_q, {sample, bins, histogram}, {ev});
+    }
+
+    return {ev, args_ev};
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:242]: In other extensions we firstly return a host event. I'd propose to keep consistency here:
```suggestion
    return {args_ev, ev};
```

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    HistType *out = static_cast<HistType *>(vout);
+
+    auto device = exec_q.get_device();
+
+    uint32_t local_size =
+        device.is_cpu()
+            ? 256
+            : device.get_info<sycl::info::device::max_work_group_size>();
+
+    uint32_t WorkPI = 128; // empirically found number
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:74]:
```suggestion
    constexpr uint32_t WorkPI = 128; // empirically found number
```

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    auto global_size = Align(CeilDiv(size, WorkPI), local_size);
+
+    auto nd_range =
+        sycl::nd_range(sycl::range<1>(global_size), sycl::range<1>(local_size));
+
+    return exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(depends);
+        uint32_t dims = 1;
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:82]:
```suggestion
        constexpr uint32_t dims = 1;
```

Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.cpp:82]: Is there a use case when `dims` might differ from 1?

File: dpnp/backend/extensions/sycl_ext/histogram.hpp [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include "utils/type_dispatch.hpp"
+#include <pybind11/pybind11.h>
+#include <sycl/sycl.hpp>
+
+namespace dpctl_td_ns = dpctl::tensor::type_dispatch;
+// using dpctl::tensor::type_dispatch::num_types;
```
Comment (antonwolfy) [inline dpnp/backend/extensions/sycl_ext/histogram.hpp:31]: Do we need to remove that?

File: dpnp/dpnp_iface_histograms.py [hunk 12]
```diff
@@ -393,48 +454,77 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+    )
+    _manager.add_event_pair(ht_ev, mem_ev)
+
+    if usm_type != n_usm_type:
+        n = dpnp.asarray(n_casted, dtype=ntype, usm_type=usm_type)
+    else:
+        n = dpnp.astype(n_casted, ntype, copy=False)
 
     if density:
         # pylint: disable=possibly-used-before-assignment
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:524]: I guess that pylint suppression can be removed now.

**Comment by oleksandr-pavlyk** [comment]
I think this is a bug:

```

In [4]: data, edges = dpt.concat([dpt.full(10**7, fill_value=2., dtype="f"), dpt.full(10**7, fill_value=1., dtype="f"), dpt.full(10**7, fill_value=4., dtype="f")]), dpt.asarray([-2,1, 2, 4], dtype="f")

In [5]: dpnp.histogram(data, edges)
Out[5]:
(array([       0, 10000000, 20000000]),
 usm_ndarray([-2.,  1.,  2.,  4.], dtype=float32))

In [6]: dpnp.histogram(data, edges, density=True)
Out[6]:
(array([0.       , 0.3333333, 0.3333333], dtype=float32),
 usm_ndarray([-2.,  1.,  2.,  4.], dtype=float32))
```

The density should be `(0, 1/3, 2/3)`, instead of `(0, 1/3, 1/3)`.

File: dpnp/dpnp_iface_histograms.py (hunk 3 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:492]: added comment

File: dpnp/dpnp_iface_histograms.py (hunk 4 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: Thanks! I've modified `_result_type` to use `map_dtype_to_device`

File: dpnp/backend/extensions/sycl_ext/histogram.cpp (hunk 5 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: done

File: dpnp/backend/extensions/sycl_ext/CMakeLists.txt (hunk 6 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/CMakeLists.txt:36]: Thanks, done.

File: dpnp/backend/extensions/sycl_ext/histogram_common.hpp (hunk 7 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram_common.hpp]: Updated.

File: dpnp/backend/extensions/sycl_ext/histogram.cpp (hunk 8 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: Thanks! I've changed the order

File: dpnp/backend/extensions/sycl_ext/histogram.cpp (hunk 9 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: done

File: dpnp/backend/extensions/sycl_ext/histogram.cpp (hunk 10 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: for histogram - no. For histogramdd - yes. But they uses the same common kernel

**Comment by AlexanderKalistratov** [comment]
@oleksandr-pavlyk it is not a bug. Numpy demonstrates the same behavior:
```
>>> import numpy
>>> data, edges = numpy.concatenate([numpy.full(10**7, fill_value=2., dtype="f"), numpy.full(10**7, fill_value=1., dtype="f"), numpy.full(10**7, fill_value=4., dtype="f")]), numpy.asarray([-2,1, 2, 4], dtype="f")
>>> numpy.histogram(data, edges)
(array([       0, 10000000, 20000000]), array([-2.,  1.,  2.,  4.], dtype=float32))
>>> numpy.histogram(data, edges, density=True)
(array([0.        , 0.33333333, 0.33333333]), array([-2.,  1.,  2.,  4.], dtype=float32))
```

File: dpnp/backend/extensions/sycl_ext/histogram.cpp (hunk 10 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: done

File: dpnp/backend/extensions/sycl_ext/histogram.hpp (hunk 11 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.hpp]: done

File: dpnp/dpnp_iface_histograms.py (hunk 12 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: removed

File: dpnp/dpnp_iface_histograms.py [hunk 13]
```diff
@@ -45,6 +45,12 @@
 
 import dpnp
 
+# pylint: disable=no-name-in-module
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:48]: It looks working for me without the muted rule
```suggestion
```

File: dpnp/dpnp_iface_histograms.py [hunk 14]
```diff
@@ -393,48 +454,79 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
-    simple_weights = False
-
-    if uniform_bins is not None and simple_weights:
-        # TODO: implement fast algorithm for equal bins
-        pass
-    else:
-        # Compute via cumulative histogram
-        if weights is None:
-            sa = dpnp.sort(a)
-            cum_n = _search_sorted_inclusive(sa, bin_edges)
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:421]: Looks `_search_sorted_inclusive` is never used now and can be removed.

File: dpnp/dpnp_iface_histograms.py [hunk 15]
```diff
@@ -306,6 +312,61 @@ def digitize(x, bins, right=False):
     return bins.size - dpnp.searchsorted(bins[::-1], x, side=side)
 
 
+def _find_supported_dtype(dt, supported):
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:315]: Could you please place the new functions implementation according to a lexicographical order in the file.

File: dpnp/dpnp_iface_histograms.py [hunk 16]
```diff
@@ -306,6 +312,61 @@ def digitize(x, bins, right=False):
...
+    ):
+        if a_bin_dtype == dpnp.complex128:
+            hist_dtype = dpnp.float64
+        elif a_bin_dtype == dpnp.float64:
+            hist_dtype = dpnp.complex128
+
+        if hist_dtype == dpnp.complex128:
+            a_bin_dtype = dpnp.float64
+        elif hist_dtype == dpnp.float64:
+            a_bin_dtype = dpnp.complex128
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:365]: Isn't that do the same like in above block 357-360?

File: dpnp/dpnp_iface_histograms.py [hunk 17]
```diff
@@ -306,6 +312,61 @@ def digitize(x, bins, right=False):
...
+def _result_type_for_device(dtype1, dtype2, device):
+    rt = dpnp.result_type(dtype1, dtype2)
+    return map_dtype_to_device(rt, device)
+
+
+def _align_dtypes(a_dtype, bins_dtype, ntype, device):
+    has_fp64 = device.has_aspect_fp64
+    a_bin_dtype = _result_type_for_device(a_dtype, bins_dtype, device)
+
+    supported_types = (dpnp.float32, dpnp.int64, numpy.uint64, dpnp.complex64)
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:335]: Wouldn't it be helpful to return a mapper function based on `ContigFactory`? Like it's done for ufunc, and then it will only need to check if we can cast the input arrays to expected matching dtype if any. And it will return the dtype of result histogram array we have to allocate.

File: dpnp/dpnp_iface_histograms.py [hunk 18]
```diff
@@ -306,6 +312,61 @@ def digitize(x, bins, right=False):
...
+
+    float_types = [dpnp.float32, dpnp.float64]
+    complex_types = [dpnp.complex64, dpnp.complex128]
+
+    a_bin_dtype = _find_supported_dtype(a_bin_dtype, supported_types)
+    hist_dtype = _find_supported_dtype(ntype, supported_types)
+    if hist_dtype == numpy.uint64:
+        hist_dtype = dpnp.int64
+
+    if (a_bin_dtype in float_types and hist_dtype in float_types) or (
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:347]: Should it be `else if`?
```suggestion
    elif (a_bin_dtype in float_types and hist_dtype in float_types) or (
```

File: dpnp/dpnp_iface_histograms.py [hunk 19]
```diff
@@ -393,48 +454,79 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+        dpnp.astype(weights, hist_dtype, order="C", copy=False)
+        if weights is not None
+        else None
+    )
+
+    # histogram implementation uses atomics, but atomics doesn't work with
+    # host usm memory
+    n_usm_type = "device" if usm_type == "host" else usm_type
+
+    n_casted = dpnp.zeros(
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:494]: Do we need to fill the memory with zeros?
```suggestion
    n_casted = dpnp.empty(
```

File: dpnp/dpnp_iface_histograms.py [hunk 20]
```diff
@@ -393,48 +454,79 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+        if weights_casted is not None
+        else None
+    )
+    n_usm = dpnp.get_usm_ndarray(n_casted)
+
+    ht_ev, mem_ev = sycl_ext.histogram(
+        a_usm,
+        bins_usm,
+        weights_usm,
+        n_usm,
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:516]: We know that `n_casted` is dpnp.ndarray, so we can use `get_array` method:
```suggestion
        n_casted.get_array(),
```

File: dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+// This file defines functions of dpnp.backend._lapack_impl extensions
+//
+//*****************************************************************************
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "histogram.hpp"
+
+PYBIND11_MODULE(_sycl_ext_impl, m)
```
Comment (oleksandr-pavlyk) [inline dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp:35]: I think we should change the name of the extension (and the folder containing the implementation) into something less generic, like `histogram`, or `bin_counting`.

File: dpnp/dpnp_iface_histograms.py [hunk 22]
```diff
@@ -393,48 +454,79 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
...
+    else:
+        n = dpnp.astype(n_casted, ntype, copy=False)
 
     if density:
-        # pylint: disable=possibly-used-before-assignment
-        db = dpnp.diff(bin_edges).astype(dpnp.default_float_type())
+        db = dpnp.diff(bin_edges).astype(
+            dpnp.default_float_type(sycl_queue=queue)
+        )
         return n / db / n.sum(), bin_edges
```
Comment (oleksandr-pavlyk) [inline dpnp/dpnp_iface_histograms.py:530]: Perhaps `n.sum()` should be replaced with `dpnp.sum(n)`.

File: dpnp/dpnp_iface_histograms.py (hunk 13 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:48]: It is working without muted rule in case module already being built.

But on CI pre-commit runs before build.

File: dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp (hunk 21 above)
Comment (antonwolfy) [reply dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp:35]: Probably it would be good to rename the extension something like `statistics` to group by functional block.
Moreover considering that new kernels will have a similar implementation approach.

File: .pre-commit-config.yaml [hunk 23]
```diff
@@ -52,7 +52,6 @@ repos:
     rev: 24.4.2
     hooks:
     -   id: black
-        args: ["--check", "--diff", "--color"]
```
Comment (oleksandr-pavlyk) [inline .pre-commit-config.yaml:55]: Just curious, was this change intentional?

File: dpnp/backend/extensions/sycl_ext/histogram.cpp [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+
+        auto dispatch_bins = [&](auto &weights) {
+            auto local_mem_size =
+                device.get_info<sycl::info::device::local_mem_size>() /
+                sizeof(T);
+            if (local_mem_size >= bins_count) {
+                uint32_t max_local_copies = local_mem_size / bins_count;
+                uint32_t local_hist_count = std::max(
+                    std::min(
+                        int(std::ceil((float(4 * local_size) / bins_count))),
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram.cpp]: done

File: dpnp/backend/extensions/sycl_ext/histogram_common.hpp [hunk 25]
```diff
@@ -0,0 +1,468 @@
...
+        return dpctl::tensor::math_utils::less_complex(lhs, rhs);
+    }
+};
+
+template <typename T>
+struct IsNan
+{
+    static bool isnan(const T &v)
+    {
+        if constexpr (std::is_floating_point<T>::value) {
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram_common.hpp]: done

File: dpnp/backend/extensions/sycl_ext/histogram_common.cpp [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+            return arr->get_queue() != exec_q;
+        });
+
+    if (unequal_queue != arrays.cend()) {
+        throw py::value_error(
+            get_name(*unequal_queue) +
+            " parameter has incompatible queue with parameter " +
+            get_name(&sample));
+    }
+
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram_common.cpp]: done

File: dpnp/backend/extensions/sycl_ext/histogram_common.cpp [hunk 27]
```diff
@@ -0,0 +1,196 @@
...
+        return "'" + name_it->second + "'";
+    };
+
+    auto unequal_queue =
+        std::find_if(arrays.cbegin(), arrays.cend(), [&](const array_ptr &arr) {
+            return arr->get_queue() != exec_q;
+        });
+
+    if (unequal_queue != arrays.cend()) {
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/histogram_common.cpp]: don't see much benefits in it. It is still one function call.

File: dpnp/dpnp_iface_histograms.py (hunk 14 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:421]: removed

File: dpnp/dpnp_iface_histograms.py (hunk 15 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: It doesn't seems that functions are sorted in other modules.
And I'm not sure that's rational behind this?

File: dpnp/dpnp_iface_histograms.py (hunk 16 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: No. On lines 357-360 we are updating `hist_dtype` and here we are updating `a_bin_dtype`.
But anyway it is removed

File: dpnp/dpnp_iface_histograms.py (hunk 17 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: It more complicated than for ufunc and elementwise.
`hist_dtype` is not directly depends on `a_bin_dtype`.
And supported types matrix is not cartesian product of `a_bin_dtype` and `hist_dtype` supported types.

So types selection is quite complex and is hard to implemented in generic way.
But anyway I've added functionality to return supported types combination to python and implement some version of types selection which is ok for histogram.

File: dpnp/dpnp_iface_histograms.py (hunk 18 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: No. `a_bin_dtype` is not depends on `hist_dtype`

File: dpnp/dpnp_iface_histograms.py (hunk 19 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:471]: We need, yes.
Histogram atomically increments elements of output array. So it must be set to zero in the beginning.

File: dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp (hunk 21 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/sycl_ext/sycl_ext_py.cpp]: Done

File: dpnp/dpnp_iface_histograms.py (hunk 22 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: Replaced

File: .pre-commit-config.yaml (hunk 23 above)
Comment (AlexanderKalistratov) [reply .pre-commit-config.yaml:55]: Yes.
If black is supplied with any of these arguments it stops reformat files.
Without these arguments black would reformat files and return error if any file was reformatted.
So it works on CI and can be run locally to automatically reformat files.

**Comment by AlexanderKalistratov** [comment]
@oleksandr-pavlyk @antonwolfy please re-review

File: dpnp/backend/extensions/statistics/histogram.cpp [hunk 28]
```diff
@@ -0,0 +1,289 @@
...
+                                     weights, nd_range, cgh);
+                }
+                else {
+                    auto edges = UncachedEdges(bins_edges, bins_count + 1, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+            };
+
+            auto dispatch_bins = [&](auto &weights) {
```
Comment (oleksandr-pavlyk) [inline dpnp/backend/extensions/statistics/histogram.cpp:156]: Looks like this could be `const auto &weights`.

File: dpnp/backend/extensions/statistics/common.hpp [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+        return dpctl::tensor::math_utils::less_complex(lhs, rhs);
+    }
+};
+
+template <typename T>
+struct IsNan
+{
+    static bool isnan(const T &v)
+    {
+        if constexpr (std::is_floating_point_v<T>) {
```
Comment (oleksandr-pavlyk) [inline dpnp/backend/extensions/statistics/common.hpp:104]:
```suggestion
        if constexpr (std::is_floating_point_v<T> || std::is_samve_v<T, sycl::half>) {
```

File: dpnp/backend/extensions/statistics/histogram_common.cpp [hunk 30]
```diff
@@ -0,0 +1,240 @@
...
+    if (non_contig_array != arrays.cend()) {
+        throw py::value_error(get_name(*non_contig_array) +
+                              " parameter is not c-contiguos");
+    }
+
+    auto check_overlaping = [&](const array_ptr &first,
+                                const array_ptr &second) {
+        const auto &overlap = dpctl::tensor::overlap::MemoryOverlap();
+
+        if (overlap(sample, histogram)) {
```
Comment (antonwolfy) [inline dpnp/backend/extensions/statistics/histogram_common.cpp:107]: I guess it should be
```suggestion
        if (overlap(first, second)) {
```

File: dpnp/backend/extensions/statistics/common.hpp (hunk 29 above)
Comment (antonwolfy) [reply dpnp/backend/extensions/statistics/common.hpp:104]: Or I guess `sycl::ext::oneapi::experimental::is_genfloat_v<...>` can be used.

File: dpnp/backend/extensions/statistics/histogram.cpp [hunk 31]
```diff
@@ -0,0 +1,289 @@
...
+#include <pybind11/stl.h>
+
+// dpctl tensor headers
+#include "dpctl4pybind11.hpp"
+#include "utils/type_dispatch.hpp"
+
+#include "histogram.hpp"
+#include "histogram_common.hpp"
+
+#include <iostream>
```
Comment (antonwolfy) [inline dpnp/backend/extensions/statistics/histogram.cpp:44]: Do we need that include or it is a kind of debug leftover?

File: dpnp/dpnp_iface_histograms.py [hunk 32]
```diff
@@ -306,6 +299,50 @@ def digitize(x, bins, right=False):
...
+    if (a_bin_dtype, ntype) in supported_types:
+        return a_bin_dtype, ntype
+
+    for sample_type, hist_type in supported_types:
+        if _can_cast(a_bin_dtype, sample_type, device) and _can_cast(
+            ntype, hist_type, device
+        ):
+            return sample_type, hist_type
+
+    return None, None
```
Comment (antonwolfy) [inline dpnp/dpnp_iface_histograms.py:343]: Could you please add a test to cover that line?

File: dpnp/dpnp_iface_histograms.py (hunk 19 above)
Comment (antonwolfy) [reply dpnp/dpnp_iface_histograms.py:471]: It would be helpful to add a comment somehwere clarifying that `statistics_ext.histogram` expects result array with zero values.

File: dpnp/backend/extensions/statistics/histogram.cpp (hunk 28 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/statistics/histogram.cpp]: done

File: dpnp/backend/extensions/statistics/common.hpp (hunk 29 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/statistics/common.hpp]: Done, thank you.
@antonwolfy while your suggestion looks good, let's for this PR stick to the version proposed by Sasha.

File: dpnp/backend/extensions/statistics/histogram_common.cpp (hunk 30 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/statistics/histogram_common.cpp]: Thanks for noticing this!
Fixed

File: dpnp/backend/extensions/statistics/histogram.cpp (hunk 31 above)
Comment (AlexanderKalistratov) [reply dpnp/backend/extensions/statistics/histogram.cpp]: Removed

File: dpnp/dpnp_iface_histograms.py [hunk 33]
```diff
@@ -306,6 +299,50 @@ def digitize(x, bins, right=False):
...
+    if dtype in [dpnp.float64, dpnp.complex128]:
+        return device.has_aspect_fp64
+
+    if dtype == dpnp.float16:
+        return device.has_aspect_fp16
+
+    return True
+
+
+def _can_cast(dtype1, dtype2, device):
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py]: Thanks!

File: dpnp/dpnp_iface_histograms.py (hunk 19 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:458]: added comment

File: dpnp/dpnp_iface_histograms.py (hunk 32 above)
Comment (AlexanderKalistratov) [reply dpnp/dpnp_iface_histograms.py:329]: I can't. This line is unreachable for current configuration of supported types

**Review by antonwolfy — APPROVED** [review]
Thank you @AlexanderKalistratov for improving the histogram implementation.
I don't have any more comment.
